### PR TITLE
Feature/use osrf images

### LINKED ISSUE #12: Docker file only works for ros_distros target at ubuntu 20.04
I'll close #11 and #9 so we can track everything here (the root problem)

### PR DESCRIPTION
**Pull request by caioaamaral**
### Description
1. Feature explanation
   - Update ros_ws to use OSRF DockerHub's images

### Related Issues:
- fix partially #12

### CODE REVIEW

File: .github/workflows/action-image-scan-deploy-dashing.yaml [hunk 1]
```diff
@@ -23,6 +21,5 @@ jobs:
         username: ${{ secrets.DOCKERHUB_USERNAME }}
         password: ${{ secrets.DOCKERHUB_TOKEN }}
     
-    - run: docker build --build-arg ROS_VERSION=$ROS_VERSION --build-arg UBUNTU_DISTRO=$UBUNTU_DISTRO --build-arg ROS_DISTRO=$ROS_DISTRO --target devel-build -t ros-$ROS_DISTRO:devel .
-    - run: docker tag ros-$ROS_DISTRO:devel openbr/ros-$ROS_DISTRO:devel
-    - run: docker push openbr/ros-$ROS_DISTRO:devel
+    - run: docker build --build-arg ROS_DISTRO=$ROS_DISTRO -t orise-robotics/ros-$ROS_DISTRO:devel .
+    - run: docker push orise-robotics/ros-$ROS_DISTRO:devel
```
Comment (mateus-amarante) [inline .github/workflows/action-image-scan-deploy-dashing.yaml]: I noticed the images in the DockerHub has the name `open-br`. Shouldn't we update any settings there?

File: Dockerfile [hunk 2]
```diff
@@ -1,67 +1,24 @@
...
-
-# install bootstrap tools
-RUN apt-get update && apt-get install --no-install-recommends -y \
-    build-essential \
-    git \
+# install ros tools
+RUN apt-get update && \
+    apt-get install --no-install-recommends -y \
     python3-colcon-mixin \
-    python3-rosdep \
```
Comment (mateus-amarante) [inline Dockerfile:37]: Why did you remove these packages?

File: Dockerfile [hunk 3]
```diff
@@ -1,67 +1,24 @@
...
-    ros-$ROS_DISTRO-ros-base \
-    nano \
-    bash-completion \
-    && rm -rf /var/lib/apt/lists/*
-
-ENV ROS_DISTRO=$ROS_DISTRO
-
-RUN echo "source /opt/ros/$ROS_DISTRO/setup.bash" >> /etc/bash.bashrc
-RUN echo "PS1='\[\033[01;35m\]ros-$ROS_DISTRO@devel\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '" >> /etc/bash.bashrc
-RUN echo "PS1='\[\033[01;35m\]ros-$ROS_DISTRO@devel\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '" >> /etc/skel/.bashrc
```
Comment (mateus-amarante) [inline Dockerfile:63]: This is not set in the best shape but I like to have a different color by default. I agree to remove this now and let this update when we target the issue https://github.com/orise-robotics/ros_ws/issues/1

File: build-images.sh [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/bin/env bash
 
-function usage() {
+usage() {
     printf "Usage: $0 [ROS_DISTRO] t]\n"
     printf "Build test and devel images for the provided ROS_DISTRO\n\n"
     printf "Options:\n"
```
Comment (mateus-amarante) [inline build-images.sh:6]: This usage() message is not compatible with the API updates

File: build-images.sh [hunk 5]
```diff
@@ -9,50 +9,63 @@ function usage() {
...
+  local SHORT=h,d:
+  local LONG=distro:
+
+  local OPTS=$(getopt -o $SHORT -l $LONG --name "$0" -- "$@")
+
+  if [ $? != 0 ] ; then echo "Failed to parse options...exiting." >&2 ; exit 1 ; fi
+
+  eval set -- "$OPTS"
+
+  . settings.sh  # set initial values
```
Comment (mateus-amarante) [inline build-images.sh]: Personally, I do not like this strategy because it does not seem to be very common and I think it can get confusing. I prefer to restrict the usage to the argument interface.

File: build-images.sh [hunk 6]
```diff
@@ -9,50 +9,63 @@ function usage() {
...
+
+  while true ; do
+    case "$1" in
+      -d | --distro )
+        ROS_DISTRO="$2"
+        shift 2
```
Comment (mateus-amarante) [inline build-images.sh:30]: I think I prefer to have the distro as a positional required argument (as it was before) to assure the right image is built. By the way, I think the scripts `run-tests.sh` and `run-devel.sh` should have the same behavior

File: build-images.sh [hunk 7]
```diff
@@ -9,50 +9,63 @@ function usage() {
...
         ;;
-    *)
+      --test )
+        BUILD_TEST="$2"
+        shift 2
+        ;;
+      -- )
+        shift
         break
```
Comment (mateus-amarante) [inline build-images.sh:38]: What is this option?

File: build-images.sh [hunk 8]
```diff
@@ -9,50 +9,63 @@ function usage() {
...
-    ;;
-*)
-    echo "Not supported ROS_DISTRO '$ROS_DISTRO'" 1>&2
+  done
+
+}
+
+validate_ros_distro() {
+  local distro=$1
+  if [[ ! $VALID_ROS_DISTROS =~ (^|[[:space:]])"${distro}"($|[[:space:]]) ]]; then
```
Comment (mateus-amarante) [inline build-images.sh:51]: Just a question. What is this verification exactly?

File: build-images.sh [hunk 9]
```diff
@@ -9,50 +9,63 @@ function usage() {
...
+}
+
+build_image() {
+  validate_ros_distro ${ROS_DISTRO}
+
+  if [[ "$BUILD_TEST" == "true" ]]; then
+    docker build --build-arg ROS_DISTRO=$ROS_DISTRO -t orise-robotics/ros-$ROS_DISTRO:test -f Dockerfile.test .
+  fi
+
+  docker build --build-arg ROS_DISTRO=$ROS_DISTRO -t orise-robotics/ros-$ROS_DISTRO:devel .
```
Comment (mateus-amarante) [inline build-images.sh]: I think we can change this behavior to build only the `devel` or the `test` images depending on the `--test` argument

File: devel_entrypoint.sh [hunk 10]
```diff
@@ -1,4 +1,5 @@
-#!/bin/env bash
+#!/usr/bin/env bash
+
 set -e
-source /opt/ros/$ROS_DISTRO/setup.bash
+source "/ros_entrypoint.sh"
```
Comment (mateus-amarante) [inline devel_entrypoint.sh]: Where is this file?

File: run-devel.sh [hunk 11]
```diff
@@ -10,7 +10,7 @@ function usage() {
     exit 0
 }
 
-ROS_DISTRO="noetic"
+. settings.sh  # set initial values
```
Comment (mateus-amarante) [inline run-devel.sh]: Personally, I do not like this strategy. Please check my other comment about this.

File: run-devel.sh [hunk 12]
```diff
@@ -31,20 +31,21 @@ while [ -n "$1" ]; do
     shift
 done
 
-CONTAINER="ros-$ROS_DISTRO-devel"
-IMAGE="ros-$ROS_DISTRO:devel"
+CONTAINER="orise-robotics_ros-$ROS_DISTRO-devel"
```
Comment (mateus-amarante) [inline run-devel.sh]: I think the container name became very big. I think we could keep the way it was

File: settings.sh [hunk 13]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+# Default options
+VALID_ROS_DISTROS="kinetic melodic noetic foxy dashing"
+ROS_DISTRO="noetic"
+VOLUMES_FOLDER=$HOME/dev_ws/volumes/
```
Comment (mateus-amarante) [inline settings.sh]: VALID_ROS_DISTROS should be a constant, not a config variable. I think ROS_DISTRO fits better as a positional argument. VOLUMES_FOLDER could also be an argument for the script, but looking at this file now I see it would be nice to have a default value fixed as the user wants...

So, now I think it is not so bad to have a settings file, especially for the VOLUMES_FOLDER option. If you like the idea of keeping it, I suggest renaming the file to something like `['build-image' | ''].[config | settings]` to indicate it is not a script.

File: run-devel.sh [hunk 14]
```diff
@@ -31,20 +31,21 @@ while [ -n "$1" ]; do
...
+IMAGE="orise-robotics/ros-$ROS_DISTRO:devel"
 
-VOLUMES_FOLDER=$HOME/dev_ws/volumes/$CONTAINER
+VOLUME="${VOLUMES_FOLDER}/$CONTAINER"
+echo "folder name ${VOLUMES_FOLDER}"
```
Comment (mateus-amarante) [inline run-devel.sh]: I think this is just for debug, right?

**Review by mateus-amarante — COMMENTED** [review]
Great job!

Just check my comments and the CI errors. Additionally, I also want to ask you to review the `README.md` file given the updates you proposed.

File: devel_entrypoint.sh (hunk 10 above)
Comment (caioaamaral) [reply devel_entrypoint.sh]: inside the container, the base image is responsible for it

Comment (mateus-amarante) [reply devel_entrypoint.sh]: I think there is no need to override the entry point of the base image (check [this](https://stackoverflow.com/questions/43237183/dockerfile-how-use-cmd-or-entrypoint-from-base-image))

File: .github/workflows/action-image-scan-deploy-dashing.yaml (hunk 1 above)
Comment (ivan1993br) [reply .github/workflows/action-image-scan-deploy-dashing.yaml]: I had to create a new one, after this PR, I will delete the old organization, but it is **oriserobotics** due to docker naming rules

File: Dockerfile [hunk 15]
```diff
@@ -1,67 +1,24 @@
-####### TEST BUILD #######
```
Comment (ivan1993br) [inline Dockerfile:1]: WHere are you installing sudo here as we discussed?

File: build-images.sh (hunk 5 above)
Comment (ivan1993br) [reply build-images.sh]: I must say that I "disagree" somehow, I agree that it is not common, it is more common using  config file passed as argument, however I think that like this is easier as it is more expansible in number of arguments, maybe for now we can pass from command line and this file?

File: build-images.sh (hunk 6 above)
Comment (ivan1993br) [reply build-images.sh:30]: But the DISTRO before wasn't optional?

File: run-devel.sh (hunk 12 above)
Comment (ivan1993br) [reply run-devel.sh]: Maybe if someone has multiple containers just ros-distro-devel can generate some conflict, how about orise-ros-distro-devel?

**Review by ivan1993br — COMMENTED** [review]
Nice! Just see my comments about the docker naming

File: build-images.sh (hunk 5 above)
Comment (mateus-amarante) [reply build-images.sh]: After seeing it closely I agree it is good to have this as a config file, I'll just change the file extension to differentiate it from the other scripts

File: build-images.sh (hunk 6 above)
Comment (mateus-amarante) [reply build-images.sh:30]: It wasn't optional for the `build-images.sh` script. I think it could remain required for this script and optional for the test and devel scripts.

File: run-devel.sh (hunk 12 above)
Comment (mateus-amarante) [reply run-devel.sh]: I prefer @ivan1993br's suggestion

File: Dockerfile (hunk 15 above)
Comment (caioaamaral) [reply Dockerfile:1]: removed here 4f064a18bc03200e75a3aac269586258d520cc7b

But still there are some issues with CI during the container scanning (only with dashing?)

File: Dockerfile (hunk 2 above)
Comment (caioaamaral) [reply Dockerfile:37]: > Why did you remove these packages?

the ros_base already comes with the [proper rosdep version](https://github.com/osrf/docker_images/blob/b37c2bedbdd06ee0d12aa6581359ad23900aa335/ros/noetic/ubuntu/focal/ros-base/Dockerfile#L8). Still, I must change the base image to ros:${ROS_DISTRO}-ros-base (I've forgot about this change).

I think is better letting the base_image install the right rosdep version rather than passing it as an argument.

File: build-images.sh (hunk 8 above)
Comment (caioaamaral) [reply build-images.sh:51]: just checking if `${distro}` is in `$VALID_ROS_DISTROS`

File: run-devel.sh (hunk 12 above)
Comment (caioaamaral) [reply run-devel.sh]: what about we remove the ros prefix? like

`orise-<distro>-devel`

Comment (mateus-amarante) [reply run-devel.sh]: sounds good too

File: .github/workflows/action-image-scan-deploy-dashing.yaml (hunk 1 above)
Comment (caioaamaral) [reply .github/workflows/action-image-scan-deploy-dashing.yaml]: changed 780dbf2

File: build-images.sh (hunk 9 above)
Comment (caioaamaral) [reply build-images.sh]: changed

File: devel_entrypoint.sh (hunk 10 above)
Comment (caioaamaral) [reply devel_entrypoint.sh]: fixed 8fbca5f

File: run-devel.sh (hunk 14 above)
Comment (caioaamaral) [reply run-devel.sh]: fixed

File: .config [hunk 16]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+# Default options
+readonly VALID_ROS_DISTROS="kinetic melodic noetic foxy dashing"
```
Comment (mateus-amarante) [inline .config:4]: Don't you think this should be in the `run-devel.sh` file. This should not be the user's responsibility.

File: build-images.sh [hunk 17]
```diff
@@ -1,6 +1,6 @@
 #!/bin/env bash
 
-function usage() {
+usage() {
     printf "Usage: $0 [ROS_DISTRO] t]\n"
     printf "Build test and devel images for the provided ROS_DISTRO\n\n"
```
Comment (mateus-amarante) [inline build-images.sh:5]: With the `-d` option, this is not valid anymore. Please, update the help message or define it back as a positional argument (Personally, I like to have this as a positional argument).

**Review by mateus-amarante — APPROVED** [review]
Great update!

Just check my last comments.

File: .config (hunk 16 above)
Comment (caioaamaral) [reply .config:4]: I prefer to have all constants in a single place, I few like this would make slightly easier making further changes. But I can move this to `run-devel.sh ` if desired.

File: build-images.sh (hunk 17 above)
Comment (caioaamaral) [reply build-images.sh:5]: Updated 0c0c59d 



> (Personally, I like to have this as a positional argument)

Seems odd to use ROS_DISTRO as positional argument here, an then use it as a flag at run-devel.sh
